### ISSUE TITLE
Unnecessary dirty check in ExtUserRegisterPosthook

### ISSUE THREAD

**Issue by alipang** [BobRay/ClassExtender#5]
I use ClassExtender for my users, but have only the minimal set of fields in the registration. It seems the classextended data only is created when there's a classextended field. If I have decided to call ExtUserRegisterPosthook, my intention is to use the fields, and no dirty check is needed (?).

I got the WSOD on a user presentation page because of this, and fixed it by adding a hidden classextended field in the registration form.

**Comment by BobRay** [comment]
Out of curiosity, why are you calling ExtUserRegisterPosthook, when there are no classextended fields in the form?

Can you tell me where the error was thrown?

**Comment by alipang** [comment]
The registration process is:

1. Enter username, email, password twice, math question. Submit.
2. Click confirmation link in email.
3. User is automatically logged in and redirected to an update form with every possible field.

If the user doesn't update, then the extra fields aren't created. If someone then looks at the user presentation page, like https://fakenews.com/user/56/, something goes wrong. I think it's in

```
$userid = (int) $_REQUEST["userid"];
$c = $modx->newQuery('userData');
$c->where(array('User.id' => $userid));
$fakeuser = $modx->getObjectGraph('userData', '{"Profile":{},"User":{}}', $c, false);
$fakeuserFields = array_merge($fakeuser->toArray(), $fakeuser->Profile->toArray());
```
Doesn't it look like a table join problem?

**Comment by BobRay** [comment]
I think you're correct, but maybe the new user has no user profile.  I'll try to find time to look into it. If you're right, I'm not sure how to fix it, since creating the fields would load up the DB with records for users who don't exist (and the "users" may be bots who register multiple times). 

If you're correct, maybe you could put a sanity check in front of the getObjectGraph() call using getObject() to get the userdata record and do something else if the return is empty.

**Comment by BobRay** [comment]
On second thought, regardless of the cause putting this after the getObjectGraph() call might fix it:

```
if ($fakeuser)  {
    $fakeuserFields = array_merge($fakeuser->toArray(), $fakeuser->Profile->toArray());
} else {
   /* return a message or put something in the error log */
}
```

**Comment by alipang** [comment]
So you mean there actually will be unreferenced data in the classextender table? Hmm, definitely a problem... I thought it always had a reference in the ordinary user related tables. Well, a note in the documentation then perhaps. I think I stick with my solution and try to remember to observe my table for unexpected growth. Looking at the unrelated might give a hint of where the world is going :-)

**Comment by BobRay** [comment]
No, I mean that there will be users in the regular users table who bailed out before any data was created in the classextender table. They are perhaps not real users, but rather people who will never be back.

Now that there's a created on field for users and the last login date field, it's easy to purge inactive or bogus users.

**Comment by alipang** [comment]
But if you purge them, then the classextended data should go away too? It shouldn't be much overhead.

**Comment by BobRay** [comment]
TBH, I don't think it will unless you add code to delete it in your utility snippet. I could be wrong -- it depends on how the related-objects are defined in the schema and how your utility snippet finds the records to be purged.  

I would back up the tables: modUser, modUserProfile, and your custom table before messing with it.

**Comment by BobRay** [comment]
Closing this as old. Reopen if it's still an issue.
